src/TokenBuyerV2.sol

```diff
@@ -0,0 +1,459 @@
+    }
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      VIEW FUNCTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /// @notice Get how much additional sellToken balance or allowance this contract needs in order to fund its current obligations plus `additionalTokens`.
+    /// @param additionalTokens an additional amount of `paymentToken` liability to use in this sellToken requirement calculation.
+    /// @return insufficientBalance the amount of additional sellToken the treasury needs
+    /// @return insufficientAllowance the amount of additional sellToken allowance this contract needs
+    function sellTokenNeeded(uint256 additionalTokens)
+        public
+        view
+        returns (uint256 insufficientBalance, uint256 insufficientAllowance)
+    {
+        uint256 paymentTokenAmount = paymentTokenAmountNeeded() + additionalTokens;
+        uint256 sellTokenAmount = sellTokenAmountPerPaymentTokenAmount(paymentTokenAmount);
+        uint256 sellTokenBalance = sellToken.balanceOf(treasury);
+        uint256 sellTokenAllowance = sellToken.allowance(treasury, address(this));
+        insufficientBalance = sellTokenAmount > sellTokenBalance ? sellTokenAmount - sellTokenBalance : 0;
+        insufficientAllowance = sellTokenAmount > sellTokenAllowance ? sellTokenAmount - sellTokenAllowance : 0;
+    }
+
+    /// @notice Returns the amount of payment tokens this contract is willing to swap
+    /// @return amount of tokens
+    function paymentTokenAmountNeeded() public view returns (uint256) {
+        IPayer _payer = payer;
+        uint256 _tokensAvailable = paymentToken.balanceOf(address(_payer));
+        uint256 totalDebt = _payer.totalDebt();
+        unchecked {
+            uint256 neededTokens = baselinePaymentTokenAmount + totalDebt;
+            if (_tokensAvailable > neededTokens) {
+                return 0;
+            }
+            return neededTokens - _tokensAvailable;
+        }
+    }
+
+    /// @notice Returns the `sellToken`/`paymentToken` price this contract is willing to swapp at, including the discount
+    /// @return The price, in 18 decimal format
+    function price() public view returns (uint256) {
+        unchecked {
+            return (priceFeed.price() * (10_000 - botDiscountBPs)) / 10_000;
+        }
+    }
+
+    /// @notice Returns the amount of sellToken this contract will send in exchange for `tokenAmount` payment tokens
+    /// @param paymentTokenAmount the amount of paymentToken tokens
+    /// @return amount of sellToken the contract will sell for `tokenAmount` of payment tokens
+    function sellTokenAmountPerPaymentTokenAmount(uint256 paymentTokenAmount) public view returns (uint256) {
+        unchecked {
+            // Example:
+            // if paymentTokenAmount == 3400000000 (3400 USDC) (6 decimals)
+            // and price() == 1745910000000000000000 (1745.91) (18 decimals)
+            // ((3400000000 * 1e36) / 1745910000000000000000) / 1e6 = 1.947408515e18 (3400/1745.91)
```

Comment:
nit: would be cool to update this example comment to show how it works when sellTokenUnit is different from 1e18